library/data/list/array_list.lean

```diff
@@ -0,0 +1,163 @@
+import init.data.quot
+universe u
+
+def defined_array.equiv (α) {n} (s : set (fin n)) (a b : array α n) := ∀ x ∈ s, a^.read x = b^.read x
+
+def defined_array.setoid (α) {n} (s : set (fin n)) : setoid (array α n) :=
+⟨defined_array.equiv α s,
+ λa x _, rfl,
+ λa b ab x h, eq.symm (ab x h),
+ λa b c ab bc x h, eq.trans (ab x h) (bc x h)⟩
+
+def defined_array (α) {n} (s : set (fin n)) := quotient (defined_array.setoid α s)
+
+def defined_array.lift_on {α n s} (q : @defined_array α n s) (β) (f : array α n → β)
+  (h : ∀ (a b : array α n), (∀ x ∈ s, a^.read x = b^.read x) → f a = f b) : β :=
+@quotient.lift _ β (defined_array.setoid α s) f h q
+
+def defined_array.read {α n s} (q : @defined_array α n s) (x) (h : x ∈ s) : α :=
+q^.lift_on (Π x ∈ s, α) (λa x _, a^.read x) (λ a b ab, funext $ λ x, funext $ λh, ab x h) x h
+
+def defined_array.write {α n s} (q : @defined_array α n s) (x) (h : x ∈ s) (v : α) : defined_array α s :=
+q^.lift_on _ (λa, @quotient.mk _ (defined_array.setoid α s) $ a^.write x v)
+  (λa b h, quot.sound (λi hi, show (if x = i then v else a^.read i) = (if x = i then v else b^.read i), by rw h i hi))
+
+def array_list.valid (size capacity) : set (fin capacity) := λ ⟨x, h⟩, x < size
+
+structure array_list (α : Type u) :=
+(capacity length : nat)
+(data : defined_array α (array_list.valid length capacity))
+(len_le : length ≤ capacity)
+
+def array_list.mk_data {α length capacity} (a : array α capacity) : defined_array α (array_list.valid length capacity) :=
+@quotient.mk _ (defined_array.setoid α (array_list.valid length capacity)) a
+
+local notation `⟦`:max a `⟧` := array_list.mk_data a
+
+def mk_array_list {α} [inhabited α] (capacity := 10) : array_list α :=
+{ capacity := capacity,
+  length := 0,
+  data := ⟦mk_array capacity (default α)⟧,
+  len_le := nat.zero_le capacity }
+
+def array.to_array_list {α n} (l : array α n) : array_list α :=
+{ capacity := n,
+  length := n,
+  data := ⟦l⟧,
+  len_le := le_refl n }
+
+def mk_array_list_of_fn {α n cap} [inhabited α] (f : fin n → α) (H : n ≤ cap) : array_list α :=
+{ capacity := cap,
+  length := n,
+  data := ⟦{data := λ⟨x, _⟩, if h : x < n then f ⟨x, h⟩ else default α}⟧,
+  len_le := H }
+
+def array.to_array_list_cap {α n cap} [inhabited α] (arr : array α n) (H : n ≤ cap) : array_list α :=
+mk_array_list_of_fn arr^.read H
+
+namespace array_list
+
+def nil {α} : array_list α :=
+{ capacity := 0,
+  length := 0,
+  data := ⟦array.nil⟧,
+  len_le := le_refl 0 }
+
+lemma lt_capacity {α} (l : array_list α) {n} (h : n < l^.length) : n < l^.capacity :=
+nat.lt_of_lt_of_le h l^.len_le
+
+def read {α} (l : array_list α) (x : fin l^.length) : α :=
+let ⟨n, h⟩ := x in l^.data^.read ⟨n, l^.lt_capacity h⟩ h
+
+def to_array {α} (l : array_list α) : array α l^.length :=
+{data := l^.read}
+
+def write {α} (l : array_list α) : fin l^.length → α → array_list α :=
+λ ⟨n, h⟩ v, { l with data := l^.data^.write ⟨n, l^.lt_capacity h⟩ h v }
+
```

Comment:
@leodemoura I tried your version and a few others:

    def write2 {α} : ∀ (l : array_list α), fin l^.length → α → array_list α 
    | ⟨cap, len, data, hle⟩ ⟨n, h⟩ v :=
      ⟨cap, len, data^.write ⟨n, nat.lt_of_lt_of_le h hle⟩ h v, hle⟩ 
    
    def write3 {α} (l : array_list α) : fin l^.length → α → array_list α
    | ⟨n, h⟩ v :=
      match l, h with
      ⟨cap, len, data, hle⟩, h := ⟨cap, len, data^.write ⟨n, nat.lt_of_lt_of_le h hle⟩ h v, hle⟩
      end

    def write4 {α} (l : array_list α) : fin l^.length → α → array_list α
    | ⟨n, h⟩ v :=
      let cap := l^.capacity, len := l^.length, data := l^.data, hle := l^.len_le in
      ⟨cap, len, data^.write ⟨n, nat.lt_of_lt_of_le h hle⟩ h v, hle⟩
    
    def write5 {α} (l : array_list α) : fin l^.length → α → array_list α
    | ⟨n, h⟩ v :=
      match l^.capacity, l^.length, h, l^.data, l^.len_le with
      cap, len, h, data, hle := ⟨cap, len, data^.write ⟨n, nat.lt_of_lt_of_le h hle⟩ h v, hle⟩
      end

    set_option trace.array.update true
    vm_eval ([1, 2, 3]^.to_array_list^.write (fin.of_nat 1) 10)^.read (fin.of_nat 1) -- non-destructive
    vm_eval ([1, 2, 3]^.to_array_list^.write2 (fin.of_nat 1) 10)^.read (fin.of_nat 1) -- destructive
    vm_eval ([1, 2, 3]^.to_array_list^.write3 (fin.of_nat 1) 10)^.read (fin.of_nat 1) -- destructive
    vm_eval ([1, 2, 3]^.to_array_list^.write4 (fin.of_nat 1) 10)^.read (fin.of_nat 1) -- non-destructive
    vm_eval ([1, 2, 3]^.to_array_list^.write5 (fin.of_nat 1) 10)^.read (fin.of_nat 1) -- destructive
    
    example {α} (l : array_list α) (n : ℕ) (h : n < l^.length) (v : α) : (l^.write ⟨n, h⟩ v)^.capacity = l^.capacity := rfl
    example {α} (l : array_list α) (n : ℕ) (h : n < l^.length) (v : α) : (l^.write2 ⟨n, h⟩ v)^.capacity = l^.capacity := rfl --failed
    example {α} (l : array_list α) (n : ℕ) (h : n < l^.length) (v : α) : (l^.write3 ⟨n, h⟩ v)^.capacity = l^.capacity := rfl --failed
    example {α} (l : array_list α) (n : ℕ) (h : n < l^.length) (v : α) : (l^.write4 ⟨n, h⟩ v)^.capacity = l^.capacity := rfl
    example {α} (l : array_list α) (n : ℕ) (h : n < l^.length) (v : α) : (l^.write5 ⟨n, h⟩ v)^.capacity = l^.capacity := rfl

Version 2 is the one you presented above, and version 3 is the suggested match expression. Both of them fail to reduce projections definitionally. I have a few alternatives, versions 4 and 5.

I assume that `let` expressions allow controlling the evaluation order, so I was hoping that version 4 would work. But it doesn't achieve the destructive write - do you know why? Perhaps it is just because `l` is still in scope, even though it is not being used. If this version can be made to work, it is a good alternative compilation for `{s with ...}`.

The only version which has both properties is version 5, which uses a match expression on the projections, but this involves an extra construction / destruction step.